library/ssl_tls13_server.c

```diff
@@ -24,16 +24,782 @@
+    ciphersuite_info = NULL;
+    for ( j = 0, p = cipher_suites_start; j < cipher_suites_len; j += 2, p += 2 )
+    {
+        for ( i = 0; cipher_suites[i] != 0; i++ )
+        {
+            if( MBEDTLS_GET_UINT16_BE(p, 0) != cipher_suites[i] )
+                continue;
+
+            ciphersuite_info = mbedtls_ssl_ciphersuite_from_id(
+                               cipher_suites[i] );
+
+            if( ciphersuite_info == NULL )
+            {
+                MBEDTLS_SSL_DEBUG_MSG(
+                1,
+                ( "mbedtls_ssl_ciphersuite_from_id: should never happen" ) );
+                return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+            }
+
+            goto have_ciphersuite;
+
+        }
+    }
+
+    return( MBEDTLS_ERR_SSL_HANDSHAKE_FAILURE );
+
+have_ciphersuite:
+
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "selected ciphersuite: %s",
+                                ciphersuite_info->name ) );
+
+    ssl->session_negotiate->ciphersuite = cipher_suites[i];
+    ssl->handshake->ciphersuite_info = ciphersuite_info;
+
+    /* List all the extensions we have received */
+#if defined(MBEDTLS_DEBUG_C)
+    ssl_tls13_debug_print_client_hello_exts( ssl );
+#endif /* MBEDTLS_DEBUG_C */
+
+    /*
+     * Determine the key exchange algorithm to use.
+     * There are three types of key exchanges supported in TLS 1.3:
+     * - (EC)DH with ECDSA,
+     * - (EC)DH with PSK,
+     * - plain PSK.
+     *
+     * The PSK-based key exchanges may additionally be used with 0-RTT.
+     *
+     * Our built-in order of preference is
+     *  1 ) Plain PSK Mode
+     *  2 ) (EC)DHE-PSK Mode
+     *  3 ) Certificate Mode
+     */
+
+    if( !ssl_tls13_check_certificate_key_exchange( ssl ) )
+    {
+        MBEDTLS_SSL_DEBUG_MSG(
+                1,
+                ( "ClientHello message misses mandatory extensions." ) );
+        MBEDTLS_SSL_PEND_FATAL_ALERT( MBEDTLS_SSL_ALERT_MSG_MISSING_EXTENSION ,
+                                      MBEDTLS_ERR_SSL_ILLEGAL_PARAMETER );
+        return( MBEDTLS_ERR_SSL_ILLEGAL_PARAMETER );
+    }
+
+    if( hrr_required == 1 )
+        return( SSL_CLIENT_HELLO_HRR_REQUIRED );
+
+    return( 0 );
+}
+
+/* Update the handshake state machine */
+
+static int ssl_tls13_postprocess_client_hello( mbedtls_ssl_context* ssl )
+{
+    int ret = 0;
+
+    ret = mbedtls_ssl_tls13_key_schedule_stage_early( ssl );
+    if( ret != 0 )
+    {
+        MBEDTLS_SSL_DEBUG_RET( 1,
+             "mbedtls_ssl_tls1_3_key_schedule_stage_early", ret );
+        return( ret );
+    }
+
+    mbedtls_ssl_handshake_set_state( ssl, MBEDTLS_SSL_SERVER_HELLO );
+    return( 0 );
+
+}
+
+/*
+ * Main entry point from the state machine; orchestrates the otherfunctions.
+ */
+
+static int ssl_tls13_process_client_hello( mbedtls_ssl_context *ssl )
+{
+
+    int ret = 0;
+    unsigned char* buf = NULL;
+    size_t buflen = 0;
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> parse client hello" ) );
+
+    ssl->major_ver = MBEDTLS_SSL_MAJOR_VERSION_3;
```

Comment:
Is it needed ? Or move ssl->minor_ver set here ?

Reply:
Actually we only need ssl->minor_ver, but consider they should show in pair, I just set the ssl->major_ver also here.

Reply:
```suggestion
    ssl->major_ver = MBEDTLS_SSL_MAJOR_VERSION_3;
    ssl->minor_ver = MBEDTLS_SSL_MINOR_VERSION_4;
```